test/bio_read_test.c

```diff
@@ -0,0 +1,253 @@
+/*
+ * Copyright 2018 The OpenSSL Project Authors. All Rights Reserved.
+ *
+ * Licensed under the OpenSSL license (the "License").  You may not use
+ * this file except in compliance with the License.  You can obtain a copy
+ * in the file LICENSE in the source distribution or at
+ * https://www.openssl.org/source/license.html
+ */
+#include <stdio.h>
+#include <string.h>
+#include <openssl/evp.h>
+#include <openssl/bio.h>
+#include <openssl/rand.h>
+#include <openssl/bio.h>
+
+#include "testutil/output.h"
+#include "testutil.h"
+
+#if defined(_WIN32) || defined(OPENSSL_NO_DGRAM)
+int setup_tests(void)
+{
+    return 1;
+}
+#else
+#include <sys/socket.h>
+#include <netinet/in.h>
+#include <unistd.h>
+
+#define PACKET_COUNT 10
+
+/*
+ * this test case opens a pair of v4 or v6 sockets, bound to localhost.
+ * A process is forked to do the writing. The write path is tested in
+ * bio_write_test case.
+ *
+ * This code exercises the BIO_read side of things.
+ *
+ * The foreground process reads 5 packets from the socket and discards
+ * them.
+ *
+ * The packets being send to the other destination are just ignored for now.
+ */
+
+static int read_socket_and_discard(int fd, int count, unsigned portnum)
+{
+  char buf[512];
+  BIO  *in;
+  int  ret = 0;
+  BIO_ADDR *ba;
+  unsigned short port;
+
+  /* do not wait forever! */
+  alarm(60);
+
+  ba = BIO_ADDR_new();
+
+  in = BIO_new_dgram(fd, BIO_CLOSE);
+
+  while(--count > 0 && (ret=BIO_read(in, buf, 512))>0) {
+
+    /* now check out the bio structure for the origin of the packet */
+    BIO_get_dgram_origin(in, ba);
+    port = ntohs(BIO_ADDR_rawport(ba));
+    if(port!=0 && port != portnum) { exit(5); }
+  }
+
+  if(ret <= 0) {
+    test_printf_stderr("BIO_read returned %d\n", ret);
+  }
+
+  BIO_ADDR_free(ba);
+
+  return count;
+}
+
+static int write_packets(BIO *bio, int count, BIO_ADDR *dst1, BIO_ADDR *dst2)
+{
+  const char hello[]="hellohellohello";
+  int toggle = 0;
+  int ret;
+
+  while(--count > 0) {
+    const int sizeofhello = sizeof(hello);
+
+    if(toggle && dst2 != NULL) {
```

Comment:
Nit: Both branches are single lines so `{}` should be omitted (if one branch was a single line and one was not we would require `{}` on both)